### DISCUSSION TITLE
How to deploy multiple editors on a custom server

### DISCUSSION THREAD

**Discussion by master034231** [mui/toolpad#2908] [Help and support]
I have three toolpad applications deployed on a custom server. I want to unify the domain name and port of the portal by deploying a nginx reverse proxy on the front-end, but I don't know how to configure it

I noticed that in the https://demo.toolpad.io/ There should be a string on it that can be used to distinguish, like this https://demo.toolpad.io/_toolpad/app/clp0z4ixk00hzooacbewtcpu2/pages/xr7laoc6 can use clp0z4ixk00hzooacbewtcpu2

How should this be set? Thank you

**Comment by Janpot** [comment]
In custom server scenarios, the end-user application is separated from the editor. At the moment the editor can only connect to a single Toolpad application. Keep in mind, on order for the application to be editable, it needs to be started in dev mode. An application in dev mode is unsafe to expose to untrusted users. In short term I could see a possibility to allow the editor to connect to multiple Toolpad applications if they are running on the same host.

https://demo.toolpad.io/ is a deprecated version of Toolpad. Currently, its multitenancy capabilities can't be replicated with a custom server.

Could you tell us a bit more about your use-case? About what you're trying to build/ What kind of applications are you building? Who will be authoring applications and who will be end-users?

**Reply by JerryWu1234** [reply]
@Janpot 
Let me provide you with an example. I have a platform where each user can separately create their own project online. All user data is stored on my cloud server. Currently, My solution that I can only provide a single cloud server for each user. **I hope to enable we can create more than one project in the Toolpad. and then one user can modify himself own directory during accessing the toolpad website  so that  we can reduce more virtual machine**
![image](https://github.com/mui/mui-toolpad/assets/27261522/9dabe519-52b5-4473-b463-6af13eeb73fc)

**Reply by JerryWu1234** [reply]
@Janpot one who is using Toolpad for him product. it's difficult step into next stage, because of this one.  please check this issue frist

**Reply by Janpot** [reply]
@JerryWu1234 What are you building exactly? A hosted version of Toolpad?

**Reply by JerryWu1234** [reply]
@Janpot 
yes.
a hosted version of Toolpad.
I have discussed with them @chendwww. @master034231, actually they are in the one tame.
such as stackblitz. an online editor.
They also want to build an online Toolpad.

**Comment by chendwww** [comment]
Is that possible to build multitenancy capabilities by Next.js App router?

**Reply by Janpot** [reply]
I've opened https://github.com/mui/mui-toolpad/issues/3012 which proposes running a Toolpad application as a server component. In the meantime, can you elaborate a bit more on the exact use-case you have in mind?

**Reply by chendwww** [reply]
Yes, We have a in hourse CRM system which shared by different sub companies and teams, we want to use Toolpad to create a page builder which could connect with our CRM data.  The current custom server (https://mui.com/toolpad/concepts/custom-server/) only can host one toolpad in a server, in this case, we have to use more than 300 servers if we want to each team of us to have their own toolpad editor. We want to have a multitenancy page builder that not cost too many servers

**Reply by Janpot** [reply]
@chendwww Trying to read between the lines here, but do I understand correctly that you are trying to build a low-code dashboard builder tailored to your own data/components? Will the end-users of this dashboard builder be programmers? I'm asking because this is a use-case we see pop up regularly, but Toolpad wasn't really created with this use-case in mind. In it's current form it may be challenging to scale to 300 apps running simultaneously. If there's sufficient demand for this use-case we may think about providing a solution for this.

**Reply by master034231** [reply]
@Janpot At present, it seems that we hope to achieve multi tenant access through this method: creating a folder at the root path of the project, which has several sub projects, and after startup, the URL parameters can determine which sub projects to access

**Reply by JerryWu1234** [reply]
@Janpot 
Could you please prioritize helping look at this issue?
Thanks.
I can pick this problem up if anything is ready.

**Reply by Janpot** [reply]
> Could you please prioritize helping look at this issue?

That would be a challenge. Hosting the editor other than your local machine isn't exactly what we envisioned Toolpad to do. Having a vite dev server over the wire will likely be too resource intensive to scale well. Running 300 of them on one machine is also something I don't see happening in the near future. To enable it we may have to rethink Toolpad architecture a bit likely also make different trade-offs in features as a result. That's why I'm interested in your use-case:

* are you trying to build a hosted version of Toolpad, e.g. as a SaaS?
* what are the capabilities of end-users of the editors? 
  * Will they build custom components? 
  * Will they only consume custom components? 
  * Will they build backend functions/data providers? 
  * Will they write javascript expressions to connect elements together? Or just use constants for property values.
  * Or are these the things that will be baked into the platform you create?
  * how do you see them deploying their final applications?
  * ...

One thing we often see is that people are trying to build a low-code dashboarding tool on top of their own platform. Where end users don't really have to be able to program, but can drag&drop, add layout and set basic configuration to a set of predefined components that are connected to a custom backend.

**Reply by master034231** [reply]
@Janpot 
Regarding the several issues you mentioned:
1. We are indeed trying to build a managed version of the toolpad
2. The end users of the editor are mostly non-technical personnel, so most users will not build custom components. In most cases, they will only directly use the components we have developed
3. We provide the backend services and provide a visual data table configuration function. We can directly allow users to connect components with the data tables we provide through simple configuration (this is also something we have been doing before, and it has been completed here)
4. Most users do not know how to write JS, so we will consider modifying the structure design of some Toolpads so that people without JS experience can quickly build their own programs
5. We still hope that users can deploy the final application on the servers we provide, and we will provide the ability for users to customize domain names (I'm not sure if this can answer your question, if not, you can add it)

**Reply by JerryWu1234** [reply]
@Janpot 
>are you trying to build a hosted version of Toolpad, e.g. as a SaaS?

they are building a SaaS base on the Toolpad.
